File: src/main/java/com/micahelias/opengl/VertexArray.java
package com.micahelias.opengl;



import static org.lwjgl.opengl.GL20.*;
import static org.lwjgl.opengl.GL30.*;

import java.util.ArrayList;


public class VertexArray {

  int m_RendererID;

  public VertexArray() {
    m_RendererID = glGenVertexArrays();
  }

  public void addBuffer(VertexBuffer vb, VertexBufferLayout layout) {

    bind();
    vb.bind();

    ArrayList<VertexBufferElement> elements = layout.getElements();
    int offset = 0;
    for (int i = 0; i < elements.size(); i++) {
      VertexBufferElement element = elements.get(i);
      glEnableVertexAttribArray(i);
      glVertexAttribPointer(i, element.count, element.type, element.normalized, layout.getStride(), offset);
      offset += element.getTypeSize();
    }

  }

  public void bind() {
    glBindVertexArray(m_RendererID);
  }

  public void unbind() {
    glBindVertexArray(0);
  }


}

File: opengl/vertex_array.go
package opengl

import (
	"github.com/go-gl/gl/v3.3-core/gl"
)

type VertexArray struct {
	m_RendererID uint32
}

func NewVertexArray() *VertexArray {
	var vao uint32
	gl.GenVertexArrays(1, &vao)
	return &VertexArray{m_RendererID: vao}

}

func (vao *VertexArray) AddBuffer(vb VertexBuffer, layout VertexBufferLayout) {
	vao.Bind()
	vb.Bind()
	elements := layout.GetElements()
	offset := 0
	for i := 0; i < len(elements); i++ {
		element := elements[i]
		gl.EnableVertexAttribArray(uint32(i))
		gl.VertexAttribPointer(uint32(i), int32(element.m_Count), uint32(element.m_Type), element.m_Normalized, int32(layout.GetStride()), gl.PtrOffset(offset))
		offset += element.GetTypeSize() * element.m_Count
	}
}

func (vao *VertexArray) Bind() {
	gl.BindVertexArray(vao.m_RendererID)
}

func (vao *VertexArray) Unbind() {
	gl.BindVertexArray(0)
}
